### PR TITLE
add Chroma

### PR DESCRIPTION
**Pull request by croquelois**
Code tested and already used for a while inside those repo by various users:
- https://github.com/croquelois/forgeChroma
- https://github.com/maybleMyers/chromaforge

The code already leverage the future approval of https://github.com/lllyasviel/huggingface_guess/pull/2 in the meantime the code works by injecting the information in the loader, it rely on this check: `chroma_is_in_huggingface_guess = hasattr(huggingface_guess.model_list, "Chroma")` to determine if the workaround is needed. The workaround will be easy to delete once the other PR is approved.

The code in `nn` is a variation of `flux.py` with changes present in https://github.com/lodestone-rock/ComfyUI_FluxMod (author of Chroma model)

## A few comment about Chroma:

Chroma is quite similar to Flux-Schnell, but the guidance part has been removed:
- `double_block`, don't have `txt_mod` nor `img_mod` anymore
- `single_blocks`, don't have `modulation` anymore
- `final_layer`, don't have `adaLN_modulation` anymore
- `guidance_in`, `time_in`, `vector_in` are removed

instead there is an independent multilayered network called `distilled_guidance_layer` which handle the same function without being deeply intertwined to the different layers

another change is on the text encoder, Flux is relying on both CLIP_L and T5, but Chroma only need T5.

last change, the original Flux model is padding the result of the T5 Tokenizer to force it to at least 256 token. Chroma is trained without the padding.

image

### CODE REVIEW

File: backend/diffusion_engine/chroma.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+from backend.args import dynamic_args
+from backend.modules.k_prediction import PredictionFlux
+from backend import memory_management
+
+class Chroma(ForgeDiffusionEngine):
+    def __init__(self, estimated_config, huggingface_components):
+        super().__init__(estimated_config, huggingface_components)
+        self.is_inpaint = False
+
+        clip = CLIP(
```
Comment (croquelois) [inline backend/diffusion_engine/chroma.py:18]: difference with Flux: only one text encoder

File: backend/diffusion_engine/chroma.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            diffusers_scheduler=None,
+            k_predictor=k_predictor,
+            config=estimated_config
+        )
+
+        self.text_processing_engine_t5 = T5TextProcessingEngine(
+            text_encoder=clip.cond_stage_model.t5xxl,
+            tokenizer=clip.tokenizer.t5xxl,
+            emphasis_name=dynamic_args['emphasis_name'],
+            min_length=1
```
Comment (croquelois) [inline backend/diffusion_engine/chroma.py:42]: difference with Flux: no padding to the encoding

File: backend/huggingface/Chroma/model_index.json [hunk 3]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (croquelois) [inline backend/huggingface/Chroma/model_index.json:1]: roughly the same than Flux with only one text encoder and one tokenizer

File: backend/huggingface/Chroma/scheduler/scheduler_config.json [hunk 4]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (croquelois) [inline backend/huggingface/Chroma/scheduler/scheduler_config.json:1]: copy&paste from flux file, same as for all the files bellow where I don't put notes

File: backend/loader.py [hunk 5]
```diff
@@ -22,9 +22,10 @@
 from backend.diffusion_engine.sdxl import StableDiffusionXL, StableDiffusionXLRefiner
 from backend.diffusion_engine.sd35 import StableDiffusion3
 from backend.diffusion_engine.flux import Flux
+from backend.diffusion_engine.chroma import Chroma
 
 
-possible_models = [StableDiffusion, StableDiffusion2, StableDiffusionXLRefiner, StableDiffusionXL, StableDiffusion3, Flux]
+possible_models = [StableDiffusion, StableDiffusion2, StableDiffusionXLRefiner, StableDiffusionXL, StableDiffusion3, Chroma, Flux]
```
Comment (croquelois) [inline backend/loader.py:28]: Chroma first because it inherit from Flux, so the `isinstance` line 572 will not confuse both

File: backend/loader.py [hunk 6]
```diff
@@ -478,14 +482,36 @@ def split_state_dict(sd, additional_state_dicts: list = None):
 
     return state_dict, guess
 
+# To be removed once PR merged on huggingface_guess
+chroma_is_in_huggingface_guess = hasattr(huggingface_guess.model_list, "Chroma")
```
Comment (croquelois) [inline backend/loader.py:486]: this work-around make it work while I wait for the huggingface_guess PR to be merged, I'll clean it up after. code was tested with and without the huggingface_guess changes.

File: backend/loader.py [hunk 7]
```diff
@@ -478,14 +482,36 @@ def split_state_dict(sd, additional_state_dicts: list = None):
...
 
+if not chroma_is_in_huggingface_guess:
```
Comment (croquelois) [inline backend/loader.py:488]: to be cleaned after huggingface_guess PR

File: backend/loader.py [hunk 8]
```diff
@@ -478,14 +482,36 @@ def split_state_dict(sd, additional_state_dicts: list = None):
...
+        }
+        unet_remove_config = ['guidance_embed']
 @torch.inference_mode()
 def forge_loader(sd, additional_state_dicts=None):
     try:
         state_dicts, estimated_config = split_state_dict(sd, additional_state_dicts=additional_state_dicts)
     except:
         raise ValueError('Failed to recognize model type!')
     
+    if not chroma_is_in_huggingface_guess \
```
Comment (croquelois) [inline backend/loader.py:504]: to be cleaned after huggingface_guess PR

File: backend/loader.py [hunk 9]
```diff
@@ -540,6 +566,8 @@ def forge_loader(sd, additional_state_dicts=None):
         else:
             huggingface_components['scheduler'].config.prediction_type = prediction_types.get(estimated_config.model_type.name, huggingface_components['scheduler'].config.prediction_type)
 
+    if not chroma_is_in_huggingface_guess and estimated_config.huggingface_repo == "Chroma":
```
Comment (croquelois) [inline backend/loader.py:569]: to be cleaned after huggingface_guess PR

File: backend/nn/chroma.py [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+from dataclasses import dataclass
+
+import math
+import torch
+
+from torch import nn
+from einops import rearrange, repeat
+from backend.attention import attention_function
+from backend.utils import fp16_fix, tensor2parameter
+from backend.nn.flux import attention, rope, timestep_embedding, EmbedND, MLPEmbedder, RMSNorm, QKNorm, SelfAttention
```
Comment (croquelois) [inline backend/nn/chroma.py:12]: the file mostly rely on Flux to avoid code duplication

File: backend/nn/chroma.py [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+
+class Approximator(nn.Module):
```
Comment (croquelois) [inline backend/nn/chroma.py:14]: difference with Flux: new class for the new guidance layer

File: backend/nn/chroma.py [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+        self.txt_attn = SelfAttention(dim=hidden_size, num_heads=num_heads, qkv_bias=qkv_bias)
+        self.txt_norm2 = nn.LayerNorm(hidden_size, elementwise_affine=False, eps=1e-6)
+        self.txt_mlp = nn.Sequential(
+            nn.Linear(hidden_size, mlp_hidden_dim, bias=True),
+            nn.GELU(approximate="tanh"),
+            nn.Linear(mlp_hidden_dim, hidden_size, bias=True),
+        )
+
+    def forward(self, img, txt, mod, pe):
+        (img_mod1, img_mod2), (txt_mod1, txt_mod2) = mod
```
Comment (croquelois) [inline backend/nn/chroma.py:59]: difference with Flux: the modulated layers are not inside the double/single stream blocks anymore, it move to the `Approximator`, and it is feed to the blocks

File: backend/nn/chroma.py [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+        x = x + mod.gate * output
+        x = fp16_fix(x)
+        return x
+
+
+class LastLayer(nn.Module):
+    def __init__(self, hidden_size, patch_size, out_channels):
+        super().__init__()
+        self.norm_final = nn.LayerNorm(hidden_size, elementwise_affine=False, eps=1e-6)
+        self.linear = nn.Linear(hidden_size, patch_size * patch_size * out_channels, bias=True)
```
Comment (croquelois) [inline backend/nn/chroma.py:127]: difference with Flux: the `adaLN_modulation` disappears replace by the `Approximator` which is feed to the forward pass through `mod`

File: backend/nn/chroma.py [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+        pe_dim = hidden_size // num_heads
+        if sum(axes_dim) != pe_dim:
+            raise ValueError(f"Got {axes_dim} but expected positional dim {pe_dim}")
+
+        self.hidden_size = hidden_size
+        self.num_heads = num_heads
+
+        self.pe_embedder = EmbedND(dim=pe_dim, theta=theta, axes_dim=axes_dim)
+        self.img_in = nn.Linear(self.in_channels, self.hidden_size, bias=True)
+        self.distilled_guidance_layer = Approximator(64, guidance_out_dim, guidance_hidden_dim, guidance_n_layers)
```
Comment (croquelois) [inline backend/nn/chroma.py:157]: difference with Flux: the new guidance layer is here

File: backend/nn/chroma.py [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+            [
+                SingleStreamBlock(self.hidden_size, self.num_heads, mlp_ratio=mlp_ratio)
+                for _ in range(depth_single_blocks)
+            ]
+        )
+
+        self.final_layer = LastLayer(self.hidden_size, 1, self.out_channels)
+        
+    @staticmethod
+    def distribute_modulations(tensor, single_block_count: int = 38, double_blocks_count: int = 19):
```
Comment (croquelois) [inline backend/nn/chroma.py:182]: difference with Flux: this method take the output of the approximator to distribute it to the different blocks + last layer

File: backend/nn/chroma.py [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+            raise ValueError("Input img and txt tensors must have 3 dimensions.")
+        img = self.img_in(img)
+        device = img.device
+        dtype = img.dtype # torch.bfloat16
+        nb_double_block = len(self.double_blocks)
+        nb_single_block = len(self.single_blocks)
+        
+        mod_index_length = nb_double_block*12 + nb_single_block*3 + 2
+        distill_timestep = timestep_embedding(timesteps.detach().clone(), 16).to(device=device, dtype=dtype)
+        distil_guidance = timestep_embedding(torch.zeros_like(timesteps), 16).to(device=device, dtype=dtype)
```
Comment (croquelois) [inline backend/nn/chroma.py:257]: difference with Flux: no distil guidance anymore, it's all zero now

File: backend/nn/chroma.py [hunk 17]
```diff
@@ -0,0 +1,307 @@
...
+        modulation_index = timestep_embedding(torch.arange(mod_index_length), 32).to(device=device, dtype=dtype)
+        modulation_index = modulation_index.unsqueeze(0).repeat(img.shape[0], 1, 1)
+        timestep_guidance = torch.cat([distill_timestep, distil_guidance], dim=1).unsqueeze(1).repeat(1, mod_index_length, 1)
+        input_vec = torch.cat([timestep_guidance, modulation_index], dim=-1)
+        mod_vectors = self.distilled_guidance_layer(input_vec)
```
Comment (croquelois) [inline backend/nn/chroma.py:262]: here we call the new guidance layer

File: backend/nn/chroma.py [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+        mod_vectors_dict = self.distribute_modulations(mod_vectors, nb_single_block, nb_double_block)
+        
+        txt = self.txt_in(txt)
+        ids = torch.cat((txt_ids, img_ids), dim=1)
+        del txt_ids, img_ids
+        pe = self.pe_embedder(ids)
+        del ids
+        for i, block in enumerate(self.double_blocks):
+            img_mod = mod_vectors_dict[f"double_blocks.{i}.img_mod.lin"]
```
Comment (croquelois) [inline backend/nn/chroma.py:271]: and here we distribute it through the blocks

File: backend/nn/chroma.py [hunk 19]
```diff
@@ -0,0 +1,307 @@
...
+        for i, block in enumerate(self.single_blocks):
+            single_mod = mod_vectors_dict[f"single_blocks.{i}.modulation.lin"]
+            img = block(img, mod=single_mod, pe=pe)
+        del pe
+        img = img[:, txt.shape[1]:, ...]
+        final_mod = mod_vectors_dict["final_layer.adaLN_modulation.1"]
+        img = self.final_layer(img, final_mod)
+        return img
+
+    def forward(self, x, timestep, context, **kwargs):
```
Comment (croquelois) [inline backend/nn/chroma.py:285]: the rest of the code is just `flux.py`

**Comment by johnnycaifan** [comment]
Is anyone else experiencing issues with inpaint? I feel like it became more unstable — now it deforms the models even when I set the weight to 0.65, whereas before I could go up to 0.8 without any problems.

**Comment by derpina-ai** [comment]
> Is anyone else experiencing issues with inpaint? I feel like it became more unstable — now it deforms the models even when I set the weight to 0.65, whereas before I could go up to 0.8 without any problems.

it was wrong before, it's correct to deform at 0.65. Check Adetailer inpainting repo for explanation on the curve of the denoise, it was badly implemented in flux
